Fix custom sidecar injection when `securityContext.RunAsUser` is supplied

jwendell opened the pull request:
Currently if the user supplies a custom `istio-proxy` container with the
`securityContext.RunAsUser` field set to a custom value, the injection
itself works but the iptables rules are applied to the wrong user (`-u`
argument of the istio-init container).
    
Plus, the init container (istio-validation) must be changed so that its `securityContext.RunAs`
fields match the proxy one.
    
This PR fixes that by changing the `-u` argument and also changing the
`uid/gid` of the init container.

istio-testing commented:
@jwendell: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790083539698454528) | true | `/test integ-pilot-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790083539698454528) | true | `/test integ-pilot-istiodremote`
integ-pilot-istiodremote-mc_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790083539379687424) | true | `/test integ-pilot-istiodremote-mc`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790083539698454528) | true | `/test integ-pilot-istiodremote`
integ-pilot-istiodremote-mc_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790083539379687424) | true | `/test integ-pilot-istiodremote-mc`
integ-telemetry-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-telemetry-istiodremote_istio/1790083545578868736) | true | `/test integ-telemetry-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett commented:
`1337` is used in a _lot_ of places as a hardcoded expectation.

- dockerfiles (including the base one)
- `istioctl`
- maybe more.

I don't object to making it be driven by install-time config, but I think it will be a little more complex.

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790083539698454528) | true | `/test integ-pilot-istiodremote`
integ-pilot-istiodremote-mc_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790083539379687424) | true | `/test integ-pilot-istiodremote-mc`
integ-telemetry-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-telemetry-istiodremote_istio/1790083545578868736) | true | `/test integ-telemetry-istiodremote`
integ-security-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-security-istiodremote_istio/1790083542215036928) | true | `/test integ-security-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

MorrisLaw commented on pkg/kube/inject/webhook.go:
```diff
@@ -568,23 +571,53 @@ func reapplyOverwrittenContainers(finalPod *corev1.Pod, originalPod *corev1.Pod,
+		return
+	}
+
+	// Locate the istio-init or istio-validation container
+	initContainer := FindContainer(InitContainerName, finalPod.Spec.InitContainers)
+	if initContainer == nil {
+		initContainer = FindContainer(ValidationContainerName, finalPod.Spec.InitContainers)
+	}
+	if initContainer == nil {
+		// should not happen
+		return
+	}
```
Wdyt about the following for readability? It's not immediately obvious why we are checking initContainer for nil twice and calling FindContainer twice until you really look at the input values and see that there are two different names. Something like the code suggestion below might make that fact more obvious:

```suggestion
        // Locate the istio-init or istio-validation container
        var initContainer *Container
        for _, name := range []string{InitContainerName, ValidationContainerName} {
            if container := FindContainer(name, finalPod.Spec.InitContainers); container != nil {
                initContainer = container
            }
        }
        if initContainer == nil {
            // should not happen
            return
        }
```

jwendell commented:
@bleggett We *already* support running the sidecar with a user/group ID other than 1337. This PR just fixes a bug in which iptables rules are misapplied when the user supplies a custom injection template.

bleggett commented:
> @bleggett We _already_ support running the sidecar with a user/group ID other than 1337. This PR just fixes a bug in which iptables rules are misapplied when the user supplies a custom injection template.

Huh, then we have a _lot_ of hardcoded references we don't need then.

bleggett commented on pkg/kube/inject/webhook.go:
```diff
@@ -568,23 +571,58 @@ func reapplyOverwrittenContainers(finalPod *corev1.Pod, originalPod *corev1.Pod,
+		return
+	}
+
+	// Make sure the validation container runs with the same uid/gid as the proxy (init container is untouched, it must run with 0)
```
```suggestion
	// Make sure the validation container runs with the same uid/gid as the proxy (init container is untouched, it must run with 0)
```

nit: not the case anymore, right?

bleggett replied:
Actually, i guess the PR desc is wrong.

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote-mc_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790083539379687424) | true | `/test integ-pilot-istiodremote-mc`
integ-security-istiodremote_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-security-istiodremote_istio/1790083542215036928) | true | `/test integ-security-istiodremote`
integ-pilot-istiodremote_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790118312886145024) | true | `/test integ-pilot-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett reviewed (approved):
LGTM, do we have any tests that actually test the overrides if you have _both_ init containers? Looks like they're all just assuming `istio-init`

jwendell commented:
> LGTM, do we have any tests that actually test the overrides if you have _both_ init containers? Looks like they're all just assuming `istio-init`

This PR has tests that cover two cases: normal and cni. In normal injection, istio-init container is used and runas must remain 0. In cni, istio-validation is used and we change its runas. It's not possible to have both init containers at the same time.

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote-mc_istio | 299f42bef6093afc2fbb23ba6213112522e1fe3c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790083539379687424) | true | `/test integ-pilot-istiodremote-mc`
integ-pilot-istiodremote_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790118312886145024) | true | `/test integ-pilot-istiodremote`
integ-security-istiodremote_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-security-istiodremote_istio/1790118315398533120) | true | `/test integ-security-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@jwendell: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote_istio/1790118312886145024) | true | `/test integ-pilot-istiodremote`
integ-security-istiodremote_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-security-istiodremote_istio/1790118315398533120) | true | `/test integ-security-istiodremote`
integ-pilot-istiodremote-mc_istio | 671c074de247eac68203a6d128e5b9eee814228c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51012/integ-pilot-istiodremote-mc_istio/1790118312512851968) | true | `/test integ-pilot-istiodremote-mc`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

costinm commented:
> > @bleggett We _already_ support running the sidecar with a user/group ID other than 1337. This PR just fixes a bug in which iptables rules are misapplied when the user supplies a custom injection template.

This is the 'we support' bot: can you define what you mean by 'support' ? Is it a documented feature with test coverage, an alpha feature with a path to stable ? 


> 
> Huh, then we have a _lot_ of hardcoded references we don't need, and fundamentally different ways to override them (which we probably also don't need)

I very much doubt every feature works when the UID is set to something else. May be worth running the cni test suite with the pods changed to run as something else to confirm. 

And please don't propose we modify CNI to look at the pod spec and use different values for each pod, the code is already far too complex and large for something so critical...

jwendell commented:
@costinm We've been running the sidecar with another uid/gid for 5 years at least I think, in OpenShift. For a few releases, we upstreamed our changes so that Istio upstream also is able to run with a different userId when running on OpenShift.

As a follow up of this PR, we should also state in the documentation (in the custom injection page) that overriding the `securityContext.RunAs` fields must be done with caution, and there are some cases where we do not honor it, such as when using tproxy mode.

@bleggett I've added a hold on this PR so you have a chance to give it another review, as I added a second commit taking care of TPROXY (also adding tests for it). Now all integration tests pass.

costinm commented:
I don't mind if OpenShift uses custom UID - I assume you have testing and support for this. 

But for most users - and admins - it is likely better to reject pods with custom gid. Can you add an option maybe - so admins can pick their poison ?

costinm commented:
For default - I am more concerned about cases when users have CNI, so maybe disallow/reject such changes when cni is used and allow it when istio init is used ?

jwendell commented:
OpenShift heavily relies on CNI, and this PR covers the CNI case as well...

bleggett commented:
> As a follow up of this PR, we should also state in the documentation (in the custom injection page) that overriding the securityContext.RunAs fields must be done with caution, and there are some cases where we do not honor it, such as when using tproxy mode.

+1

> For default - I am more concerned about cases when users have CNI, so maybe disallow/reject such changes when cni is used and allow it when istio init is used ?

We already do not allow the `istio-cni` daemonset to have a configurable UID/GID - it's hardcoded to 0.

And using init containers is mutually exclusive with using `istio-cni` (for both sidecar and ambient) - so we should be fine.

bleggett commented:
Removing DNM because 

- Tests are passing
- This is a bugfix

I do agree we should look more closely at the support status and desired UX for this feature tho.

costinm commented:
On Tue, May 14, 2024, 09:22 Ben Leggett ***@***.***> wrote:

> As a follow up of this PR, we should also state in the documentation (in
> the custom injection page) that overriding the securityContext.RunAs fields
> must be done with caution, and there are some cases where we do not honor
> it, such as when using tproxy mode.
>
> +1
>
> For default - I am more concerned about cases when users have CNI, so
> maybe disallow/reject such changes when cni is used and allow it when istio
> init is used ?
>
> We already do not allow the istio-cni daemonset to have a configurable
> UID/GID - it's hardcoded to 0.
>

The risk is istio-proxy running as gid=1234 and missing some rules that cni
sets with gid checks ( for sidecar )


And using init containers is mutually exclusive with using istio-cni (for
> both sidecar and ambient) - so we should be fine.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/51012#issuecomment-2110641003>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2SBUMIU7RYSTQMXWW3ZCI23LAVCNFSM6AAAAABHUWK6YCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJQGY2DCMBQGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bleggett commented:
> The risk is istio-proxy running as gid=1234 and missing some rules that cni sets with gid checks ( for sidecar )

Yep - that is (effectively) broken/breakable in current-state tho, if I understand it correctly. Can you raise an issue to track that?

istio-testing commented:
In response to a cherrypick label: new pull request created: #51045

bleggett commented:
> The risk is istio-proxy running as gid=1234 and missing some rules that cni sets with gid checks ( for sidecar )

I had forgotten - we actually _already_ consult the podspec in `istio-cni` when deciding what gid/uid to use - this should be fine.
